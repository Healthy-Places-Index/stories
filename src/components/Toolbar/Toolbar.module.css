.button {
  width: 30px;
  height: 30px;
  background-color: white;
  position: relative;
  border-radius: 5px;
  box-shadow: 0 0 0 2px rgb(0 0 0 / 10%);
  font-size: 14px;
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  margin-top: 5px;
  border: none;
}

.button:hover {
  background-color: #f1f1f1;
}

.button.active{
  background-color: #ccc;
}

.toolbar {
  position: 'absolute';
  left: 15px;
  top: 235px;
  display: 'flex';
  flex-direction: column;
}
