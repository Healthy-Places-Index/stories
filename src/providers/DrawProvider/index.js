import React, { useContext, createContext, useReducer, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import { useQuery, useMutation, gql } from '@apollo/client';
import { last } from 'lodash';
import { DrawPolygonMode, DrawLineStringMode, EditingMode, DrawPointMode } from 'react-map-gl-draw';

const StateContext = createContext();
const DispatchContext = createContext();

const MODES = [
  { id: 'drawPolyline', text: 'Draw Polyline', Handler: DrawLineStringMode },
  { id: 'drawPolygon', text: 'Draw Polygon', Handler: DrawPolygonMode },
  { id: 'editing', text: 'Edit Feature', Handler: EditingMode },
  { id: 'drawPoint', text: 'Edit Feature', Handler: DrawPointMode },
];

const INITIAL_STATE = {
  mode: null,
  features: [],
  ids: [],
  selectedFeatureIndex: '',
  clickRadius: 12,
  slide: null,
};

const GET_ANNOTATIONS = gql`
  query GetAnnotations($slide: ID!) {
    Slide(where: { id: $slide }) {
      id
      annotations {
        id
        title
        feature
        index
      }
    }
  }
`;

const CREATE_ANNOTATION = gql`
  mutation CreateAnnotation(
    $title: String
    $index: Int
    $feature: String
    $slide: SlideRelateToOneInput
  ) {
    createAnnotation(data: { title: $title, index: $index, feature: $feature, slide: $slide }) {
      id
    }
  }
`;

const UPDATE_ANNOTATION_FEATURE = gql`
  mutation UpdateAnnotation($id: ID!, $feature: String) {
    updateAnnotation(id: $id, data: { feature: $feature }) {
      id
    }
  }
`;

function reducer(state, [type, payload]) {
  // console.log(type, payload);
  switch (type) {
    case 'SET_MODE': {
      const modeId = payload;
      const mode = MODES.find(m => m.id === modeId);
      const modeHandler = mode ? new mode.Handler() : null;
      return { ...state, mode: modeHandler };
    }
    case 'SET_FEATURES': {
      return { ...state, features: payload };
    }
    case 'SET_IDS': {
      return { ...state, ids: payload };
    }
    case 'SET_SELECTED_FEATURE_INDEX': {
      return { ...state, selectedFeatureIndex: payload };
    }
    case 'DEL_FEATURE': {
      return {
        ...state,
        features: state.features.filter((feature, index) => payload !== index),
      };
    }
    case 'SLIDE': {
      return {
        ...state,
        slide: payload,
      };
    }
    default: {
      throw new Error(`Unhandled action type: ${type}`);
    }
  }
}

function DrawProvider({ children }) {
  const [state, dispatch] = useReducer(reducer, INITIAL_STATE);
  const [createAnnotation] = useMutation(CREATE_ANNOTATION);
  const [updateAnnotation] = useMutation(UPDATE_ANNOTATION_FEATURE);
  const { data: annotations } = useQuery(GET_ANNOTATIONS, { variables: { slide: state.slide } });

  const updateTimer = useRef();

  function onSelect(e) {
    const { selectedFeatureIndex } = e;
    dispatch(['SET_SELECTED_FEATURE_INDEX', selectedFeatureIndex]);
  }

  function onUpdate(e) {
    const { data, editType } = e;

    dispatch(['SET_FEATURES', data]);

    switch (editType) {
      case 'addFeature': {
        dispatch(['SET_MODE', 'editing']);
        const newFeature = last(data);
        const index = data.length - 1;
        createAnnotation({
          variables: {
            title: `Feature ${index}`,
            feature: JSON.stringify(newFeature),
            index,
            slide: {
              connect: {
                id: state.slide,
              },
            },
          },
          refetchQueries: [{ query: GET_ANNOTATIONS, variables: { slide: state.slide } }],
        });
        break;
      }
      case 'movePosition': {
        const { ids, selectedFeatureIndex } = state;
        const id = ids[selectedFeatureIndex];
        const feature = JSON.stringify(data[selectedFeatureIndex]);
        clearTimeout(updateTimer.current);
        updateTimer.current = setTimeout(
          () =>
            updateAnnotation({
              variables: { id, feature },
              refetchQueries: [{ query: GET_ANNOTATIONS, variables: { slide: state.slide } }],
            }),
          500
        );
        break;
      }
      default:
        break;
    }
  }

  useEffect(() => {
    if (annotations) {
      dispatch(['SET_FEATURES', annotations.Slide.annotations.map(a => JSON.parse(a.feature))]);
      dispatch(['SET_IDS', annotations.Slide.annotations.map(a => a.id)]);
    }
  }, [annotations]);

  return (
    <StateContext.Provider value={{ ...state, onUpdate, onSelect }}>
      <DispatchContext.Provider value={dispatch}>{children}</DispatchContext.Provider>
    </StateContext.Provider>
  );
}

function useDrawState() {
  const context = useContext(StateContext);

  if (context === undefined) {
    throw new Error('useDrawState must be used within a DrawProvider');
  }

  return context;
}

function useDrawDispatch() {
  const context = useContext(DispatchContext);

  if (context === undefined) {
    throw new Error('useDrawDispatch must be used within a DrawProvider');
  }

  return context;
}

function useDraw() {
  return [useDrawState(), useDrawDispatch()];
}

DrawProvider.propTypes = {
  children: PropTypes.string.isRequired,
};

export { DrawProvider, useDrawState, useDrawDispatch, useDraw };
