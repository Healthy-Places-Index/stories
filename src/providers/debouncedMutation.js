const debouncedMutation = ({ slide, timerRef, mutation, values, typename }) => {
  // eslint-disable-next-line no-underscore-dangle
  const __typename = typename || 'Slide';
  clearTimeout(timerRef.current);
  return setTimeout(
    () =>
      mutation({
        variables: {
          slide,
          ...values,
        },
        optimisticResponse: {
          __typename: 'Mutation',
          updateSlide: {
            __typename,
            id: slide,
            ...values,
          },
        },
      }),
    200
  );
};

export default debouncedMutation;
