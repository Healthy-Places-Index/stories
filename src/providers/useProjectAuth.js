import axios from 'axios';

export default async ({ req, project }) => {
  if (!req.user) return 401;
  const {
    data: {
      data: {
        Project: { user },
      },
    },
  } = await axios.post(`${req.protocol}://${req.get('Host')}/admin/api`, {
    query: `query GetProjectUser($project: ID!) {
        Project(where: { id: $project }) {
          user {
            id
          }
        }
      }
    `,
    variables: {
      project,
    },
  });

  let statusCode = null;
  if (user.id !== req.user.id) statusCode = 403;

  return statusCode;
};
