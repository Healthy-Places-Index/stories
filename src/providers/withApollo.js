/* eslint-disable react/display-name */
/* eslint-disable react/prop-types */
import React from 'react';
import withApollo from 'next-with-apollo';
import { ApolloProvider, ApolloClient, InMemoryCache } from '@apollo/client';

export default withApollo(
  ({ headers }) => {
    return new ApolloClient({
      uri: '/admin/api',
      cache: new InMemoryCache(),
      request: operation => {
        operation.setContext({
          fetchOptions: {
            credentials: 'include',
          },
          headers: { cookie: headers && headers.cookie },
        });
      },
    });
  },
  {
    render: ({ Page, props }) => {
      return (
        <ApolloProvider client={props.apollo}>
          <Page {...props} />
        </ApolloProvider>
      );
    },
  }
);
