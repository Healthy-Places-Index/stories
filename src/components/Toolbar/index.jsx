import React from 'react';

import { useDraw } from '../../providers/DrawProvider';
import styles from './Toolbar.module.css';

export default function Toolbar() {
  const [, dispatch] = useDraw();
  return (
    <div
      style={{ position: 'absolute', left: 15, top: 235, display: 'flex', flexDirection: 'column' }}
    >
      <button
        className={`${styles.button} `}
        type="button"
        tabIndex={-1}
        id="toolbar"
        onClick={() => dispatch(['SET_MODE', 'drawPolyline'])}
      >
        line
      </button>
      <button
        className={`${styles.button} `}
        type="button"
        tabIndex={-1}
        id="toolbar"
        onClick={() => dispatch(['SET_MODE', 'drawPolygon'])}
      >
        poly
      </button>
      <button
        className={`${styles.button} `}
        type="button"
        tabIndex={-1}
        id="toolbar"
        onClick={() => dispatch(['SET_MODE', 'drawPoint'])}
      >
        point
      </button>
    </div>
  );
}
