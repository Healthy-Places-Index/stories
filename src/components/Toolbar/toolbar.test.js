import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { MockedProvider } from '@apollo/client/testing';
import { DrawProvider } from '../../providers/DrawProvider';

import Toolbar from './index';

describe('Draw Toolbar', () => {
  it('matches snapshot', () => {
    const component = renderer.create(
      <MockedProvider>
        <DrawProvider>
          <Toolbar />
        </DrawProvider>
      </MockedProvider>
    );

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot('default');
  });

  it('expands on click', () => {
    const component = renderer.create(
      <MockedProvider>
        <DrawProvider>
          <Toolbar />
        </DrawProvider>
      </MockedProvider>
    );

    const button = component.root.findByType('button');
    act(() => button.props.onClick());
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot('editing');
  });
});
