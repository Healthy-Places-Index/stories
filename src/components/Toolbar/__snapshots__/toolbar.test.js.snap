// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Draw Toolbar expands on click: editing 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "left": 15,
      "position": "absolute",
      "top": 235,
    }
  }
>
  <button
    className="button"
    id="toolbar"
    onClick={[Function]}
    tabIndex={-1}
    type="button"
  >
    <img
      alt="Toggle edit mode"
      className="icon"
      src="/img/pen-active.svg"
    />
  </button>
  <button
    className="button"
    id="toolbar"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": null,
      }
    }
    tabIndex={-1}
    type="button"
  >
    <img
      alt="Draw line"
      className="icon"
      src="/img/LineString.svg"
    />
  </button>
  <button
    className="button"
    id="toolbar"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": null,
      }
    }
    tabIndex={-1}
    type="button"
  >
    <img
      alt="Draw polygon"
      className="icon"
      src="/img/Polygon.svg"
    />
  </button>
  <button
    className="button"
    id="toolbar"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": null,
      }
    }
    tabIndex={-1}
    type="button"
  >
    <img
      alt="Draw point"
      className="icon"
      src="/img/Point.svg"
    />
  </button>
</div>
`;

exports[`Draw Toolbar matches snapshot: default 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "left": 15,
      "position": "absolute",
      "top": 235,
    }
  }
>
  <button
    className="button"
    id="toolbar"
    onClick={[Function]}
    tabIndex={-1}
    type="button"
  >
    <img
      alt="Toggle edit mode"
      className="icon"
      src="/img/pen.svg"
    />
  </button>
</div>
`;
